README.md
[![build](https://github.com/osmlab/ohauth/workflows/build/badge.svg)](https://github.com/osmlab/ohauth/actions?query=workflow%3A%22build%22)
[![npm version](https://badge.fury.io/js/ohauth.svg)](https://badge.fury.io/js/ohauth)


## ohauth

A most-of-the-way OAuth 1.0 client implementation in Javascript. Meant to be
an improvement over the [default linked one](http://oauth.googlecode.com/svn/code/javascript/)
because this uses idiomatic Javascript.

If you use this on a server [different from the one authenticated against](http://en.wikipedia.org/wiki/Same_origin_policy),
you'll need to [enable](http://enable-cors.org/) and use [CORS](http://en.wikipedia.org/wiki/Cross-origin_resource_sharing)
for cross-origin resources. CORS is not available in IE before version IE10.


### Demo

Try it out at:  http://osmlab.github.io/ohauth/


### Usage

As a file

    wget https://raw.github.com/osmlab/ohauth/gh-pages/ohauth.js

With browserify

    npm install ohauth
    var ohauth = require('ohauth');


### Compatibility

* [OpenStreetMap](http://www.openstreetmap.org/) full & tested with iD
* GitHub - partial, full flow is not possible because `access_token` API is not CORS-enabled


### API

```js
// generates an oauth-friendly timestamp
ohauth.timestamp();

// generates an oauth-friendly nonce
ohauth.nonce();

// generate a signature for an oauth request
ohauth.signature("myOauthSecret", "myTokenSecret", "percent&encoded&base&string");

// make an oauth request.
ohauth.xhr(method, url, auth, data, options, callback);

// options can be a header like
{ header: { 'Content-Type': 'text/xml' } }

ohauth.xhr('POST', url, o, null, {}, function(xhr) {
    // xmlhttprequest object
});

// generate a querystring from an object
ohauth.qsString({ foo: 'bar' });
// foo=bar

// generate an object from a querystring
ohauth.stringQs('foo=bar');
// { foo: 'bar' }
```


#### Just generating the headers

```js
// create a function holding configuration
var auth = ohauth.headerGenerator({
    consumer_key: '...',
    consumer_secret: '...'
});

// pass just the data to produce the OAuth header with optional
// POST data (as long as it'll be form-urlencoded in the request)
var header = auth('GET', 'http://.../?a=1&b=2', { c: 3, d: 4 });
// or pass the POST data as an form-urlencoded
var header = auth('GET', 'http://.../?a=1&b=2', 'c=3&d=4');
```


CHANGELOG.md
:warning: = Breaking change

## 1.0.1
##### 2021-Mau-24
* remove xtend, to avoid deprecation warnings

## 1.0.0
##### 2016-Oct-21
* Upgrade all dependencies
* xhr should return an XMLHttpRequest (#14)

## 0.2.2
##### 2016-Aug-01
* Bump xtend for updated object-keys

## 0.2.1
##### 2013-Apr-27
* Make headerGenerator use token secret (#6)

## 0.2.0
##### 2013-Apr-22
* Add headerGenerator function

## 0.1.4
##### 2013-Apr-17
* Update jshashes dependency

## 0.1.3
##### 2013-Mar-26
* Use standalone flag for browserify

## 0.1.2
##### 2013-Mar-25
* Use jshashes library instead of including sha.js

## 0.1.1
##### 2013-Mar-23
* Fix regression in rawxhr call

## 0.1.0
##### 2013-Mar-13
* Initial module release


test/ohauth.js
'use strict';

var test = require('tap').test;
var ohauth = require('../.');

test('ohauth', function(t) {
    t.test('#qsString', function(t) {
        t.test('turns an object into a querystring', function(t) {
            t.deepEqual(ohauth.qsString({ foo: 1 }), 'foo=1');
            t.end();
        });
        t.test('escapes special characters', function(t) {
            t.deepEqual(ohauth.qsString({ '!\'*()': '!\'*()' }), '%21%27%2A%28%29=%21%27%2A%28%29');
            t.end();
        });
        t.end();
    });

    t.test('#stringQs', function(t) {
        t.test('turns a querystring into an object', function(t) {
            t.deepEqual(ohauth.stringQs('foo=1'), { foo: 1 });
            t.end();
        });
        t.test('handles special characters', function(t) {
            t.deepEqual(ohauth.stringQs('%21%27%2A%28%29=%21%27%2A%28%29'), { '!\'*()': '!\'*()' });
            t.end();
        });
        t.test('handles querystrings with empty arguments', function(t) {
            t.deepEqual(ohauth.stringQs(''), {});
            t.deepEqual(ohauth.stringQs('foo=1&'), { foo: 1 });
            t.end();
        });
        t.end();
    });

    t.test('#nonce', function(t) {
        t.test('generates a 6-character nonce', function(t) {
            t.deepEqual(ohauth.nonce().length, 6);
            t.end();
        });
        t.end();
    });

    t.test('#authHeader', function(t) {
        t.test('encodes an auth header', function(t) {
            t.deepEqual(ohauth.authHeader({ foo: 'bar' }), 'foo="bar"');
            t.end();
        });
        t.end();
    });

    t.test('#timestamp', function(t) {
        t.test('generates a numeric timestamp', function(t) {
            t.type(ohauth.timestamp(), 'number');
            t.end();
        });
        t.test('generates an integer timestamp', function(t) {
            t.deepEqual(ohauth.timestamp() % 1, 0);
            t.end();
        });
        t.end();
    });

    t.test('#percentEncode', function(t) {
        t.test('encodes spaces', function(t) {
            t.deepEqual(ohauth.percentEncode('foo bar'), 'foo%20bar');
            t.end();
        });
        t.end();
    });

    t.test('#headerGenerator', function(t) {
        t.test('generates a header function', function(t) {
            t.type(ohauth.headerGenerator({}), 'function');
            t.type(ohauth.headerGenerator({})('GET', 'http://foo.com/'), 'string');
            t.end();
        });
        t.end();
    });

    t.end();
});


test/index.test.js
'use strict';

var test = require('tap').test;
var ohauth = require('../.');

test('ohauth', function(t) {
  t.test('#qsString', function(t) {
    t.test('turns an object into a querystring', function(t) {
      t.same(ohauth.qsString({ foo: 1 }), 'foo=1');
      t.end();
    });
    t.test('escapes special characters', function(t) {
      t.same(ohauth.qsString({ '!\'*()': '!\'*()' }), '%21%27%2A%28%29=%21%27%2A%28%29');
      t.end();
    });
    t.end();
  });

  t.test('#stringQs', function(t) {
    t.test('turns a querystring into an object', function(t) {
      t.same(ohauth.stringQs('foo=1'), { foo: 1 });
      t.end();
    });
    t.test('handles special characters', function(t) {
      t.same(ohauth.stringQs('%21%27%2A%28%29=%21%27%2A%28%29'), { '!\'*()': '!\'*()' });
      t.end();
    });
    t.test('handles querystrings with empty arguments', function(t) {
      t.same(ohauth.stringQs(''), {});
      t.same(ohauth.stringQs('foo=1&'), { foo: 1 });
      t.end();
    });
    t.end();
  });

  t.test('#nonce', function(t) {
    t.test('generates a 6-character nonce', function(t) {
      t.same(ohauth.nonce().length, 6);
      t.end();
    });
    t.end();
  });

  t.test('#authHeader', function(t) {
    t.test('encodes an auth header', function(t) {
      t.same(ohauth.authHeader({ foo: 'bar' }), 'foo="bar"');
      t.end();
    });
    t.end();
  });

  t.test('#timestamp', function(t) {
    t.test('generates a numeric timestamp', function(t) {
      t.type(ohauth.timestamp(), 'number');
      t.end();
    });
    t.test('generates an integer timestamp', function(t) {
      t.same(ohauth.timestamp() % 1, 0);
      t.end();
    });
    t.end();
  });

  t.test('#percentEncode', function(t) {
    t.test('encodes spaces', function(t) {
      t.same(ohauth.percentEncode('foo bar'), 'foo%20bar');
      t.end();
    });
    t.end();
  });

  t.test('#headerGenerator', function(t) {
    t.test('generates a header function', function(t) {
      t.type(ohauth.headerGenerator({}), 'function');
      t.type(ohauth.headerGenerator({})('GET', 'http://foo.com/'), 'string');
      t.end();
    });
    t.end();
  });

  t.end();
});


sha.js
/*
 * A JavaScript implementation of the Secure Hash Algorithm, SHA-1, as defined
 * in FIPS PUB 180-1
 * Version 2.1a Copyright Paul Johnston 2000 - 2002.
 * Other contributors: Greg Holt, Andrew Kepert, Ydnar, Lostinet
 * Distributed under the BSD License
 * See http://pajhome.org.uk/crypt/md5 for details.
 */

function sha1() {

    /*
     * Configurable variables. You may need to tweak these to be compatible with
     * the server-side, but the defaults work in most cases.
     */
    var hexcase = 0;  /* hex output format. 0 - lowercase; 1 - uppercase        */
    var b64pad  = "="; /* base-64 pad character. "=" for strict RFC compliance   */
    var chrsz   = 8;  /* bits per input character. 8 - ASCII; 16 - Unicode      */

    /*
     * These are the functions you'll usually want to call
     * They take string arguments and return either hex or base-64 encoded strings
     */
    function hex_sha1(s){return binb2hex(core_sha1(str2binb(s),s.length * chrsz));}
    function b64_sha1(s){return binb2b64(core_sha1(str2binb(s),s.length * chrsz));}
    function str_sha1(s){return binb2str(core_sha1(str2binb(s),s.length * chrsz));}
    function hex_hmac_sha1(key, data){ return binb2hex(core_hmac_sha1(key, data));}
    function b64_hmac_sha1(key, data){ return binb2b64(core_hmac_sha1(key, data));}
    function str_hmac_sha1(key, data){ return binb2str(core_hmac_sha1(key, data));}

    /*
     * Perform a simple self-test to see if the VM is working
     */
    function sha1_vm_test() {
      return hex_sha1("abc") == "a9993e364706816aba3e25717850c26c9cd0d89d";
    }

    /*
     * Calculate the SHA-1 of an array of big-endian words, and a bit length
     */
    function core_sha1(x, len) {
      /* append padding */
      x[len >> 5] |= 0x80 << (24 - len % 32);
      x[((len + 64 >> 9) << 4) + 15] = len;

      var w = Array(80);
      var a =  1732584193;
      var b = -271733879;
      var c = -1732584194;
      var d =  271733878;
      var e = -1009589776;

      for(var i = 0; i < x.length; i += 16) {
        var olda = a;
        var oldb = b;
        var oldc = c;
        var oldd = d;
        var olde = e;

        for(var j = 0; j < 80; j++) {
          if(j < 16) w[j] = x[i + j];
          else w[j] = rol(w[j-3] ^ w[j-8] ^ w[j-14] ^ w[j-16], 1);
          var t = safe_add(safe_add(rol(a, 5), sha1_ft(j, b, c, d)),
                           safe_add(safe_add(e, w[j]), sha1_kt(j)));
          e = d;
          d = c;
          c = rol(b, 30);
          b = a;
          a = t;
        }

        a = safe_add(a, olda);
        b = safe_add(b, oldb);
        c = safe_add(c, oldc);
        d = safe_add(d, oldd);
        e = safe_add(e, olde);
      }
      return Array(a, b, c, d, e);

    }

    /*
     * Perform the appropriate triplet combination function for the current
     * iteration
     */
    function sha1_ft(t, b, c, d) {
      if(t < 20) return (b & c) | ((~b) & d);
      if(t < 40) return b ^ c ^ d;
      if(t < 60) return (b & c) | (b & d) | (c & d);
      return b ^ c ^ d;
    }

    /*
     * Determine the appropriate additive constant for the current iteration
     */
    function sha1_kt(t) {
      return (t < 20) ?  1518500249 : (t < 40) ?  1859775393 :
             (t < 60) ? -1894007588 : -899497514;
    }

    /*
     * Calculate the HMAC-SHA1 of a key and some data
     */
    function core_hmac_sha1(key, data) {
      var bkey = str2binb(key);
      if(bkey.length > 16) bkey = core_sha1(bkey, key.length * chrsz);

      var ipad = Array(16), opad = Array(16);
      for(var i = 0; i < 16; i++) {
        ipad[i] = bkey[i] ^ 0x36363636;
        opad[i] = bkey[i] ^ 0x5C5C5C5C;
      }

      var hash = core_sha1(ipad.concat(str2binb(data)), 512 + data.length * chrsz);
      return core_sha1(opad.concat(hash), 512 + 160);
    }

    /*
     * Add integers, wrapping at 2^32. This uses 16-bit operations internally
     * to work around bugs in some JS interpreters.
     */
    function safe_add(x, y) {
      var lsw = (x & 0xFFFF) + (y & 0xFFFF);
      var msw = (x >> 16) + (y >> 16) + (lsw >> 16);
      return (msw << 16) | (lsw & 0xFFFF);
    }

    /*
     * Bitwise rotate a 32-bit number to the left.
     */
    function rol(num, cnt) {
      return (num << cnt) | (num >>> (32 - cnt));
    }

    /*
     * Convert an 8-bit or 16-bit string to an array of big-endian words
     * In 8-bit function, characters >255 have their hi-byte silently ignored.
     */
    function str2binb(str) {
      var bin = Array();
      var mask = (1 << chrsz) - 1;
      for(var i = 0; i < str.length * chrsz; i += chrsz)
        bin[i>>5] |= (str.charCodeAt(i / chrsz) & mask) << (32 - chrsz - i%32);
      return bin;
    }

    /*
     * Convert an array of big-endian words to a string
     */
    function binb2str(bin) {
      var str = "";
      var mask = (1 << chrsz) - 1;
      for(var i = 0; i < bin.length * 32; i += chrsz)
        str += String.fromCharCode((bin[i>>5] >>> (32 - chrsz - i%32)) & mask);
      return str;
    }

    /*
     * Convert an array of big-endian words to a hex string.
     */
    function binb2hex(binarray) {
      var hex_tab = hexcase ? "0123456789ABCDEF" : "0123456789abcdef";
      var str = "";
      for(var i = 0; i < binarray.length * 4; i++) {
        str += hex_tab.charAt((binarray[i>>2] >> ((3 - i%4)*8+4)) & 0xF) +
               hex_tab.charAt((binarray[i>>2] >> ((3 - i%4)*8  )) & 0xF);
      }
      return str;
    }

    /*
     * Convert an array of big-endian words to a base-64 string
     */
    function binb2b64(binarray) {
      var tab = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
      var str = "";
      for(var i = 0; i < binarray.length * 4; i += 3) {
        var triplet = (((binarray[i   >> 2] >> 8 * (3 -  i   %4)) & 0xFF) << 16)
                    | (((binarray[i+1 >> 2] >> 8 * (3 - (i+1)%4)) & 0xFF) << 8 )
                    |  ((binarray[i+2 >> 2] >> 8 * (3 - (i+2)%4)) & 0xFF);
        for(var j = 0; j < 4; j++) {
          if(i * 8 + j * 6 > binarray.length * 32) str += b64pad;
          else str += tab.charAt((triplet >> 6*(3-j)) & 0x3F);
        }
      }
      return str;
    }

    return { b64_hmac_sha1: b64_hmac_sha1 };
}

if (typeof module !== 'undefined') module.exports = sha1;


index.js
'use strict';

var hashes = require('jshashes'),
    sha1 = new hashes.SHA1();


// # xtend
var hasOwnProperty = Object.prototype.hasOwnProperty;
function xtend() {
    var target = {};
    for (var i = 0; i < arguments.length; i++) {
        var source = arguments[i];
        for (var key in source) {
            if (hasOwnProperty.call(source, key)) {
                target[key] = source[key];
            }
        }
    }
    return target;
}


var ohauth = {};

ohauth.qsString = function(obj) {
    return Object.keys(obj).sort().map(function(key) {
        return ohauth.percentEncode(key) + '=' +
            ohauth.percentEncode(obj[key]);
    }).join('&');
};

ohauth.stringQs = function(str) {
    return str.split('&').filter(function (pair) {
        return pair !== '';
    }).reduce(function(obj, pair){
        var parts = pair.split('=');
        obj[decodeURIComponent(parts[0])] = (null === parts[1]) ?
            '' : decodeURIComponent(parts[1]);
        return obj;
    }, {});
};

ohauth.rawxhr = function(method, url, data, headers, callback) {
    var xhr = new XMLHttpRequest(),
        twoHundred = /^20\d$/;
    xhr.onreadystatechange = function() {
        if (4 === xhr.readyState && 0 !== xhr.status) {
            if (twoHundred.test(xhr.status)) callback(null, xhr);
            else return callback(xhr, null);
        }
    };
    xhr.onerror = function(e) { return callback(e, null); };
    xhr.open(method, url, true);
    for (var h in headers) xhr.setRequestHeader(h, headers[h]);
    xhr.send(data);
    return xhr;
};

ohauth.xhr = function(method, url, auth, data, options, callback) {
    var headers = (options && options.header) || {
        'Content-Type': 'application/x-www-form-urlencoded'
    };
    headers.Authorization = 'OAuth ' + ohauth.authHeader(auth);
    return ohauth.rawxhr(method, url, data, headers, callback);
};

ohauth.nonce = function() {
    for (var o = ''; o.length < 6;) {
        o += '0123456789ABCDEFGHIJKLMNOPQRSTUVWXTZabcdefghiklmnopqrstuvwxyz'[Math.floor(Math.random() * 61)];
    }
    return o;
};

ohauth.authHeader = function(obj) {
    return Object.keys(obj).sort().map(function(key) {
        return encodeURIComponent(key) + '="' + encodeURIComponent(obj[key]) + '"';
    }).join(', ');
};

ohauth.timestamp = function() { return ~~((+new Date()) / 1000); };

ohauth.percentEncode = function(s) {
    return encodeURIComponent(s)
        .replace(/\!/g, '%21').replace(/\'/g, '%27')
        .replace(/\*/g, '%2A').replace(/\(/g, '%28').replace(/\)/g, '%29');
};

ohauth.baseString = function(method, url, params) {
    if (params.oauth_signature) delete params.oauth_signature;
    return [
        method,
        ohauth.percentEncode(url),
        ohauth.percentEncode(ohauth.qsString(params))].join('&');
};

ohauth.signature = function(oauth_secret, token_secret, baseString) {
    return sha1.b64_hmac(
        ohauth.percentEncode(oauth_secret) + '&' +
        ohauth.percentEncode(token_secret),
        baseString);
};

/**
 * Takes an options object for configuration (consumer_key,
 * consumer_secret, version, signature_method, token, token_secret)
 * and returns a function that generates the Authorization header
 * for given data.
 *
 * The returned function takes these parameters:
 * - method: GET/POST/...
 * - uri: full URI with protocol, port, path and query string
 * - extra_params: any extra parameters (that are passed in the POST data),
 *   can be an object or a from-urlencoded string.
 *
 * Returned function returns full OAuth header with "OAuth" string in it.
 */

ohauth.headerGenerator = function(options) {
    options = options || {};
    var consumer_key = options.consumer_key || '',
        consumer_secret = options.consumer_secret || '',
        signature_method = options.signature_method || 'HMAC-SHA1',
        version = options.version || '1.0',
        token = options.token || '',
        token_secret = options.token_secret || '';

    return function(method, uri, extra_params) {
        method = method.toUpperCase();
        if (typeof extra_params === 'string' && extra_params.length > 0) {
            extra_params = ohauth.stringQs(extra_params);
        }

        var uri_parts = uri.split('?', 2),
        base_uri = uri_parts[0];

        var query_params = uri_parts.length === 2 ?
            ohauth.stringQs(uri_parts[1]) : {};

        var oauth_params = {
            oauth_consumer_key: consumer_key,
            oauth_signature_method: signature_method,
            oauth_version: version,
            oauth_timestamp: ohauth.timestamp(),
            oauth_nonce: ohauth.nonce()
        };

        if (token) oauth_params.oauth_token = token;

        var all_params = xtend({}, oauth_params, query_params, extra_params),
            base_str = ohauth.baseString(method, base_uri, all_params);

        oauth_params.oauth_signature = ohauth.signature(consumer_secret, token_secret, base_str);

        return 'OAuth ' + ohauth.authHeader(oauth_params);
    };
};

module.exports = ohauth;
